cpp/src/Slice/Grammar.y

```diff
@@ -2089,6 +2098,10 @@ tagged_type
         type = new Optional(TypePtr::dynamicCast($2));
         unit->error("only optional types can be tagged");
     }
+    if (type->usesClasses())
```

Comment:
Maybe we should have separate error messages for class types, and types that use classes. So rather than output

```
TaggedMembers.ice:112: type '::Ice::AnyClass' cannot be tagged, as it uses classes
```

We can have

```
TaggedMembers.ice:112: type '::Ice::AnyClass' class cannot be tagged
```

Reply:
Yeah, having better error messages is definitely a good thing!
But I'll probably change it to:
`class '::Test::Foo' cannot be tagged`
if the type being tagged is directly a class, and then keep the current error message for if it uses classes indirectly.

Reply:
I would also rework the "uses classes" term - it's an internal API, not a terminology meaningful to the user. Something like:
```
cannot use `AnyClass' as a tagged data member
cannot use class `FooBar' as a tagged parameter
cannot use `FooBar` as a tagged parameter because it holds directly or indirectly a class data member
```

Reply:
I can drop the 'class' in front for when it's type 'AnyClass' instead of a concrete class. I agree that is more correct.

But I think I prefer my `class 'Foo' cannot be tagged` to your first 2 suggested alternatives. It's more clear to the user that the issue is 'classes cannot be tagged', instead of making it seem like something specific to either parameters or data members. And it isn't even that 'using' it is the issue, just the act of 'tagging' it is erroneous.

I've tried thinking for alternatives to my second error message, but the best I can come up with is:
`'Foo' cannot be tagged, as it contains classes`
I think 'contains' is definitely better than 'uses'... But all the other ones I've thought up, and the one you proposed, seem way too wordy to me.

Reply:
Good point, I've spent too long in the compiler where we tie the tag to the type, and not the member...
That being said, how about:
`member 'myparam' cannot be tagged, since it's type 'AnyClass' is a class`
`member 'myparam' cannot be tagged, since it's type 'Foo' is a class`
`member 'myparam' cannot be tagged, since it's type 'Foo' contains a class`
It makes it clear that it's the member that's being tagged, but also makes it clear that the underlying reason is that it's type contains classes one way or another.

Reply:
Can you output "data member", "parameter", "return type member" as appropriate instead of always "member"?

Reply:
Yeah, it's definitely possible, it'll just take some extra code to check what the enclosing container is.